program-runtime/src/invoke_context.rs

```diff
@@ -796,75 +805,50 @@ impl<'a> InvokeContext<'a> {
         self.sysvar_cache
     }
 
+    // Should alignment be enforced during user pointer translation
+    pub fn get_check_aligned(&self) -> bool {
+        self.transaction_context
+            .get_current_instruction_context()
+            .and_then(|instruction_context| {
+                instruction_context.try_borrow_last_program_account(self.transaction_context)
+            })
+            .map(|program_account| *program_account.get_owner() != bpf_loader_deprecated::id())
+            .unwrap_or(true)
+    }
+
+    // Set should type size be checked during user pointer translation
+    pub fn get_check_size(&self) -> bool {
+        self.feature_set
+            .is_active(&check_slice_translation_size::id())
+    }
+
     // Set this instruction syscall context
     pub fn set_syscall_context(
         &mut self,
-        check_aligned: bool,
-        check_size: bool,
-        orig_account_lengths: Vec<usize>,
-        allocator: Rc<RefCell<dyn Alloc>>,
+        syscall_context: SyscallContext,
     ) -> Result<(), InstructionError> {
         *self
             .syscall_context
             .last_mut()
-            .ok_or(InstructionError::CallDepth)? = Some(SyscallContext {
-            check_aligned,
-            check_size,
-            orig_account_lengths,
-            allocator,
-        });
+            .ok_or(InstructionError::CallDepth)? = Some(syscall_context);
         Ok(())
     }
 
-    // Should alignment be enforced during user pointer translation
-    pub fn get_check_aligned(&self) -> bool {
-        self.syscall_context
-            .last()
-            .and_then(|context| context.as_ref())
-            .map(|context| context.check_aligned)
-            .unwrap_or(true)
-    }
-
-    // Set should type size be checked during user pointer translation
-    pub fn get_check_size(&self) -> bool {
-        self.syscall_context
-            .last()
-            .and_then(|context| context.as_ref())
-            .map(|context| context.check_size)
-            .unwrap_or(true)
-    }
-
-    /// Get the original account lengths
-    pub fn get_orig_account_lengths(&self) -> Result<&[usize], InstructionError> {
+    // Get this instruction's SyscallContext
+    pub fn get_syscall_context(&self) -> Result<&SyscallContext, InstructionError> {
```

Comment:
En passant not asking you to change this if you don't want because we get it wrong
in many other places, but rust accessors are supposed to be `thing()` and
`thing_mut()` eg `syscall_context()` and `syscall_context_mut()`

Reply:
right, the rest is wrong as well unfortunately, so for consistencies sake